File: WebsiteCode&DesktopAppCode/M&S/subscription.php
<?php
session_start();


include('header.php');

// Check if user is logged in
if (!isset($_SESSION["usertoken"])) {
    header("Location: login.php");
    exit();
}

// Assuming you have a database connection established
require_once "connection.php";

// Retrieve user ID from session
$userId = $_SESSION["usertoken"];
$firstName = $_SESSION["firstName"];
// Query to get user's balance, status, and end_subscription
$getSubscriptionQuery = "SELECT credits, status, end_subscription FROM userinfo WHERE id = '$userId'";
$result = mysqli_query($conn, $getSubscriptionQuery);
$balance = 0; // Default value
$status = "inactive"; // Default status
$endSubscription = null; // Default end_subscription value
if (mysqli_num_rows($result) > 0) {
    $row = mysqli_fetch_assoc($result);
    $balance = $row["credits"];
    $status = $row["status"];
    $endSubscription = $row["end_subscription"];
}

// Close the database connection
mysqli_close($conn);
?>

<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Subscription</title>

<style>



/* subscription.css */
/* subscription.css */

:root {
    --citrine: hsl(57, 97%, 45%);
    --citrine-dark: hsl(57, 97%, 35%); /* Define a darker shade of citrine */
}

body {
    font-family: Arial, sans-serif;
    margin: 0;
    padding: 0;
    background-color: #f0f0f0;
}

.container {
    max-width: 350px;
    margin: 100px auto; /* Centering the container vertically */
    padding: 20px;
    background-color: #fff;
    border-radius: 8px;
    box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
    box-sizing: border-box;
}

p {
    font-size: 18px;
    margin-bottom: 20px;
}

button {
    display: block; /* Change the display property to block */
    width: 300px; /* Set a specific width for the buttons */
    padding: 8px 16px; /* Adjust padding */
    font-size: 14px; /* Decrease font size */
    border: none;
    background-color: var(--citrine); /* Use CSS variable for background color */
    color: #000;
    cursor: pointer;
    border-radius: 4px;
    margin-bottom: 10px; /* Add margin between buttons */
    transition: background-color 0.3s; /* Add transition for smooth hover effect */
}

button:disabled {
    background-color: #ccc;
    cursor: not-allowed;
}

button:hover {
    background-color: var(--citrine-dark); /* Change to a darker shade of citrine on hover */
}


</style>




</head>
<body>


<div class="container">
<p>Welcome, <?php echo $firstName; ?>!</p>
  <p>Your current balance is: <?php echo $balance; ?> coins</p>

<p>
   Currently state is, <?php echo ucfirst($status); ?>
</p>




<!-- Button for Payment -->
<button id="payment">Payment</button>

<button class="subscribe-btn" data-subscription-option="daily_deduction">Subscribe with Daily Deduction</button>
<button class="subscribe-btn" data-subscription-option="one_month" >Subscribe for One Month Directly</button>


<!-- Button to end subscription -->
<button id="endSubscription" <?php if ($status === 'inactive' || $endSubscription !== null) echo 'disabled'; ?>>End Subscription Today</button>


<button onclick="window.location.href='index.php'">Back</button>

</div>


<script>
document.addEventListener('DOMContentLoaded', function() {
    // Event listener for Payment button
    document.getElementById('payment').addEventListener('click', function() {
        // Open add_balance.php in a new window
        const newWindow = window.open('add_balance.php', '_blank');
    });

    // Function to handle subscription confirmation
    function confirmSubscription(option) {
        if (option === 'daily_deduction' && <?php echo $balance; ?> < 1) {
            alert("Insufficient balance to subscribe for one day. Please add more coins.");
        } else if (option === 'one_month' && <?php echo $balance; ?> < 650) {
            alert("Insufficient balance. You need 650 coins.");
        } else {
            // Display a confirmation alert before redirecting to subscribe_action.php
            if (confirm("Are you sure you want to subscribe? \n \n End Subscription Today Only Available in Daily Deduction")) {
                // Redirect to subscribe_action.php with the chosen subscription option
                window.location.href = `subscribe_action.php?confirm=true&subscription_option=${option}`;
            }
        }
    }

    // Event listener for Subscribe buttons
    const subscribeButtons = document.querySelectorAll('.subscribe-btn');
    subscribeButtons.forEach(function(button) {
        button.addEventListener('click', function() {
            // Get the subscription option from the button's data attribute
            const subscriptionOption = button.getAttribute('data-subscription-option');
            
            // Check if a subscription option is provided
            if (subscriptionOption) {
                // Check if user is already subscribed
                if ("<?php echo $status; ?>" === 'active') {
                    alert("You are already subscribed.");
                } else {
                    // Call the confirmSubscription function with the chosen option
                    confirmSubscription(subscriptionOption);
                }
            } else {
                alert("Subscription option not provided.");
            }
        });
    });

    // Event listener for End Subscription button
    document.getElementById('endSubscription').addEventListener('click', function() {
        // Check if user is inactive or end_subscription is not null
        if ("<?php echo $status; ?>" === 'inactive' ) {
            alert("You are already inactive or subscription has already ended.");
        } else {
            // Confirm if the user wants to end the subscription today
            if (confirm("Are you sure you want to end your subscription today?")) {
                // Perform action to end subscription
                // For now, just redirecting to end_subscription.php
                window.location.href = 'end_subscription.php';
            }
        }
    });

});
</script>


</body>
</html>

File: WebsiteCode&DesktopAppCode/M&S/subscription.py
import datetime
import math
import pymysql

# Function to connect to the database
def connect_to_database():
    try:
        conn = pymysql.connect(host='localhost', user='root', password='', database='Movies&Series')
        return conn
    except pymysql.MySQLError as e:
        print(f"Error connecting to database: {e}")
        return None

# Function to deduct coins and update user status
def deduct_coins(user_id, start_subscription, credits, deduction_time, current_datetime, set_inactive=False):
    if deduction_time is None:
        deduction_time = start_subscription

    time_since_last_deduction = (current_datetime - deduction_time).total_seconds() / 3600
    coins_to_deduct = math.ceil(time_since_last_deduction)

    new_credits = credits - coins_to_deduct
    if new_credits < 24:
        status = 'inactive'
    else:
        status = 'inactive' if set_inactive else 'active'

    update_balance_query = (f"UPDATE userinfo SET credits = {new_credits}, deduction_time = '{current_datetime}', "
                            f"status = '{status}', end_subscription_today = 0 WHERE id = {user_id}")
    return update_balance_query, status, coins_to_deduct

# Function to check and update subscription status for users with end_subscription
def check_subscription_status(cursor, current_datetime):
    try:
        query = "SELECT id, start_subscription, end_subscription FROM userinfo WHERE status = 'active' AND end_subscription IS NOT NULL"
        cursor.execute(query)
        active_users = cursor.fetchall()

        users_affected = 0

        for user in active_users:
            user_id, start_subscription, end_subscription = user
            
            if current_datetime > end_subscription:
                update_query = "UPDATE userinfo SET status = 'inactive', end_subscription = NULL, deduction_time = NULL WHERE id = %s"
                cursor.execute(update_query, (user_id,))
                print(f"User {user_id} subscription has expired. Status updated to 'inactive'.")
                users_affected += 1

        if users_affected == 0:
            print("No one month users were affected.")
        else:
            print(f"{users_affected} one month subscription user(s) were updated to 'inactive'.")

    except Exception as e:
        print("Error:", e)

# Main function to handle the overall logic
def main():
    conn = connect_to_database()
    if conn is None:
        return

    cursor = conn.cursor()
    current_datetime = datetime.datetime.now()
    deductions_made = False

    # Check and update subscription status for users with end_subscription
    check_subscription_status(cursor, current_datetime)

    # Deduct coins for users with daily deduction enabled and end_subscription_today set to 1
    query = "SELECT id, start_subscription, credits, deduction_time FROM userinfo WHERE status = 'active' AND end_subscription IS NULL AND end_subscription_today = 1"
    cursor.execute(query)
    active_subscriptions = cursor.fetchall()

    for user_id, start_subscription, credits, deduction_time in active_subscriptions:
        update_balance_query, status, coins_to_deduct = deduct_coins(user_id, start_subscription, credits, deduction_time, current_datetime, set_inactive=True)
        try:
            cursor.execute(update_balance_query)
            conn.commit()
            print(f"Deducted {coins_to_deduct} coins from user ID {user_id}: Success")
            if status == 'inactive':
                print(f"User ID {user_id} set to inactive.")
            deductions_made = True
        except Exception as e:
            print(f"Deduction for user ID {user_id}: Failed - {str(e)}")

    # Deduct coins for users with daily deduction enabled and end_subscription_today set to 0
    query = "SELECT id, start_subscription, credits, deduction_time FROM userinfo WHERE status = 'active' AND end_subscription IS NULL AND end_subscription_today = 0"
    cursor.execute(query)
    active_subscriptions = cursor.fetchall()

    for user_id, start_subscription, credits, deduction_time in active_subscriptions:
        update_balance_query, status, coins_to_deduct = deduct_coins(user_id, start_subscription, credits, deduction_time, current_datetime, set_inactive=False)
        try:
            cursor.execute(update_balance_query)
            conn.commit()
            print(f"Deducted {coins_to_deduct} coins from user ID {user_id}: Success")
            if status == 'inactive':
                print(f"User ID {user_id} set to inactive.")
            deductions_made = True
        except Exception as e:
            print(f"Deduction for user ID {user_id}: Failed - {str(e)}")

    if not deductions_made:
        print("No deduction happened.")

    cursor.close()
    conn.close()

if __name__ == "__main__":
    main()
